veps/sig-storage/incremental-backup.md

```diff
@@ -0,0 +1,362 @@
+      <driver name='qemu' type='qcow2' cache='none' error_policy='stop' discard='unmap'/>
+      <source file='/run/kubevirt-private/libvirt/qemu/swtpm/datavolumedisk2.qcow2' index='4'>
+        <dataStore type='block'>
+          <format type='raw'/>
+          <source dev='/dev/datavolumedisk2' index='5'/>
+        </dataStore>
+      </source>
+      ...
+    </disk>
+```
+
+In the case the field is not present or set to `false`, the VM state PVC will deleted if it is not required by other features. The QCOW2 images will be deleted if exists. And the domain XML will be modified back to use the raw images as the disks images.
+
+
+### VirtualMachineBackup CRD
+
+`VirtualMachineBackup` is a namespace-scoped Custom Resource that initiates the backup process.
+
+The CR name should be a unique identifier for the backup within the namespace, with only one backup allowed per VM at a time.
+The CR spec includes the following fields:
+- `source`<br>
+Specifying the VM source for the backup.
+- `mode`<br>
+Optional. Should be either `push` or `pull`. Initially, as mentioned, only `push` will be allowed. If not specified, `push` will be the default behavior.
+- `pvcName` <br>
+Optional. A name for a PVC that the user provides that in case of `push` will store the backup output, and for `pull` will store the scratch space. If provided no PVC will be created in the process.
+- `incremental` <br>
+Optional. If not specified, a full backup will be taken. When set, the field value should be the checkpoint name from which the incremental backup should be taken.
+- `checkpointName` <br?>
+Optional. If not specified a name will be generated from a prefix of the domain name, followed by the timestamp of the backup. Example: `my-vm-checkpoint-2025-03-03T16:13:28Z`
+
+The CR status includes the following fields:
+- `checkpointName`<br>
+Updated by the controller with the name of the created Checkpoint for the current backup.
+- `pvcName` <br>
+Present only for push mode backup. Updated by the controller with the name of the PVC containing the backup output.
+- `endpoint` <br>
+Present only for pull mode backup. Update by the controller with the endpoint info from which to read the backup.
+- `conditions`<br>
+Indicates the state of the backup, such as Initializing, ReadyToPull, Failed or Deleting
+
+Examples:
+
+_Full Backup_
+```yaml
+apiVersion: kubevirt.io/v1
+kind: VirtualMachineBackup
+metadata:
+    name: backup1
+    namespace: ns1
+spec:
+    source:
+        apiGroup: kubevirt.io
+        kind: VirtualMachine
+        name: my-vm
+    mode: push
+status:
+    checkpointName: backup1-16210527022025
+    pvcName: backup-my-vm-16210527022025
+```
+
+_Incremental Backup_
+```yaml
+apiVersion: kubevirt.io/v1
+kind: VirtualMachineBackup
+metadata:
+    name: backup2
+    namespace: ns1
+spec:
+    source:
+        apiGroup: kubevirt.io
+        kind: VirtualMachine
+        name: my-vm
+    incremental: backup1-16210527022025
```

Comment:
so the user is responsible for naming checkpoints? Could his be racy? What if the incremental already exists?

Maybe we can come up with a scheme where the user gives a prefix and the checkpoint creation time is appended?

Reply:
for incremental: "checkpoint-name", this name is the checkpoint from previous backup.
For the checkpointname that should be created for the current backup I wrote:
```
- `checkpointName` <br?>
Optional. If not specified a name will be generated from a prefix of the domain name, followed by the timestamp of the backup. Example: `my-vm-checkpoint-2025-03-03T16:13:28Z`
```

Reply:
Okay, when I first read this I did not notice that `spec.incremental` and `status. checkpointName` were different.